benchmark.go

```diff
@@ -39,8 +42,102 @@ import (
 var (
 	errNegativeDuration = errors.New("duration cannot be negative")
 	errNegativeMaxReqs  = errors.New("max requests cannot be negative")
+
+	// using a global _quantiles slice and _percentileMap mainly for ease of testing, and not passing the same array
+	// around to multiple functions
+	_quantiles     = []float64{0.5, 0.9, 0.95, 0.99, 0.999, 0.9995, 1.0}
+	_percentileMap = map[float64]string{
+		0.5:    "P50",
+		0.9:    "P90",
+		0.95:   "P95",
+		0.99:   "P99",
+		0.999:  "P999",
+		0.9995: "P9995",
+		1.0:    "P100",
+	}
 )
 
+// BenchmarkParameters holds values of all benchmark parameters
+type BenchmarkParameters struct {
+	CPUs        int    `json:"cpus"`
+	Connections int    `json:"connections"`
+	Concurrency int    `json:"concurrency"`
+	MaxRequests int    `json:"maxRequests"`
+	MaxDuration string `json:"maxDuration"`
+	MaxRPS      int    `json:"maxRPS"`
+}
+
+// latencyElement stores individual quantile -> latency value mapping
+type latencyElement struct {
+	quantile string
+	latency  string
+}
+
+// LatencyMap stores latency quantile to value mappings, using a slice to preserve specific ordering of elements for serialization.
+type LatencyMap []latencyElement
+
+// Summary stores the benchmarking summary
+type Summary struct {
+	ElapsedTime   string  `json:"elapsedTime"`
+	TotalRequests int     `json:"totalRequests"`
+	RPS           float64 `json:"rps"`
+}
+
+// BenchmarkOutput stores benchmark settings and results for JSON output
+type BenchmarkOutput struct {
+	BenchmarkParameters BenchmarkParameters `json:"benchmarkParameters"`
+	Latencies           LatencyMap          `json:"latencies"`
+	Summary             Summary             `json:"summary"`
+}
+
+// MarshalJSON implements the json.Marshaler interface for the custom LatencyMap
+func (latencyMap LatencyMap) MarshalJSON() ([]byte, error) {
+	var buf bytes.Buffer
+
+	buf.WriteString("{")
+	for i, kv := range latencyMap {
```

Comment:
JSON serialization keeps the output deterministic by sorting the keys, so if we had key names that were sorted (e.g., `0.5000` sorts correctly against `0.9995`, we wouldn't need this custom marshalling

cc @jaricftw